coderd/authzquery/template.go

```diff
@@ -0,0 +1,319 @@
+		}
+		checked[v.TemplateID.UUID] = true
+	}
+
+	return versions, nil
+}
+
+func (q *AuthzQuerier) GetTemplateVersionsByTemplateID(ctx context.Context, arg database.GetTemplateVersionsByTemplateIDParams) ([]database.TemplateVersion, error) {
+	// An actor can read template versions if they can read the related template.
+	template, err := q.database.GetTemplateByID(ctx, arg.TemplateID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := q.authorizeContext(ctx, rbac.ActionRead, template); err != nil {
+		return nil, err
+	}
+
+	return q.database.GetTemplateVersionsByTemplateID(ctx, arg)
+}
+
+func (q *AuthzQuerier) GetTemplateVersionsCreatedAfter(ctx context.Context, createdAt time.Time) ([]database.TemplateVersion, error) {
+	// An actor can read execute this query if they can read all templates.
+	fetchRelated := func(tvs []database.TemplateVersion, _ time.Time) (rbac.Objecter, error) {
+		return rbac.ResourceTemplate.All(), nil
+	}
+	return authorizedQueryWithRelated(
+		q.logger,
+		q.authorizer,
+		rbac.ActionRead,
+		fetchRelated,
+		q.database.GetTemplateVersionsCreatedAfter,
+	)(ctx, createdAt)
+}
+
+func (q *AuthzQuerier) GetAuthorizedTemplates(ctx context.Context, arg database.GetTemplatesWithFilterParams, _ rbac.PreparedAuthorized) ([]database.Template, error) {
+	// TODO Delete this function, all GetTemplates should be authorized. For now just call getTemplates on the authz querier.
+	return q.GetTemplatesWithFilter(ctx, arg)
+}
+
+func (q *AuthzQuerier) GetTemplatesWithFilter(ctx context.Context, arg database.GetTemplatesWithFilterParams) ([]database.Template, error) {
+	prep, err := prepareSQLFilter(ctx, q.authorizer, rbac.ActionRead, rbac.ResourceTemplate.Type)
+	if err != nil {
+		return nil, xerrors.Errorf("(dev error) prepare sql filter: %w", err)
+	}
+	return q.database.GetAuthorizedTemplates(ctx, arg, prep)
+}
+
+func (q *AuthzQuerier) InsertTemplate(ctx context.Context, arg database.InsertTemplateParams) (database.Template, error) {
+	obj := rbac.ResourceTemplate.InOrg(arg.OrganizationID)
+	return authorizedInsertWithReturn(q.logger, q.authorizer, rbac.ActionCreate, obj, q.database.InsertTemplate)(ctx, arg)
+}
+
+func (q *AuthzQuerier) InsertTemplateVersion(ctx context.Context, arg database.InsertTemplateVersionParams) (database.TemplateVersion, error) {
+	if !arg.TemplateID.Valid {
+		// Making a new template version is the same permission as creating a new template.
+		err := q.authorizeContext(ctx, rbac.ActionCreate, rbac.ResourceTemplate.InOrg(arg.OrganizationID))
+		if err != nil {
+			return database.TemplateVersion{}, err
+		}
+	} else {
+		// Must do an authorized fetch to prevent leaking template ids this way.
+		tpl, err := q.GetTemplateByID(ctx, arg.TemplateID.UUID)
+		if err != nil {
+			return database.TemplateVersion{}, err
+		}
+		// Check the create permission on the template.
+		err = q.authorizeContext(ctx, rbac.ActionCreate, tpl)
+		if err != nil {
+			return database.TemplateVersion{}, err
+		}
+	}
```

Comment:
it would feel more consistent if you encapsulated the logic of querying the DB for the template or using `rbac.ResourceTemplate.InOrg(arg.OrganizationID)` into a `fetchRelated` function, then called `authorizedQueryWithRelated`

Reply:
Fair. We were on the fence if manual, or using the generic functions was easier to read in many cases